Refactor update list v2

jwakre opened the pull request:
This PR refactors the create_update_list function so that it no longer depends on the FROM file list. The old create_update_list implementation is kept as a fallback to be used when the TO MoM header does not include a minversion header field.

Important note about testing:

This PR also changes the testlib to create MoMs with non-zero minversion header fields. As a result, most tests will use the refactored create_update_list function.

jwakre commented on src/swupd.h:
```diff
@@ -241,44 +241,40 @@ static inline int bsearch_file_helper(const void *A, const void *B)
+	swupd_stats[5]++;
 };
 
 static inline void account_delta_miss(void)
```
This one looks like dead code

jwakre commented on src/manifest.c:
```diff
@@ -165,6 +166,14 @@ static struct manifest *manifest_from_file(int version, char *component, bool he
 		}
+
+		if (strncmp(line, "minversion:", 11) == 0) {
+			minversion = strtoull(c, NULL, 10);
```
I'll fix this

jwakre commented on src/stats.c:
```diff
@@ -28,17 +28,16 @@
 #include "swupd.h"
 
-int swupd_stats[8];
+int swupd_stats[7];
```
Ok, sure.

otaviobp commented on src/swupd.h:
```diff
@@ -241,44 +241,40 @@ static inline int bsearch_file_helper(const void *A, const void *B)
+	swupd_stats[4]++;
 };
 
 static inline void account_delta_hit(void)
```
"Do it it this?" > I have not idea what I was trying to ask.

I just did a grep for this function on your branch and it's not being called.

jwakre replied:
Can you double check? When I run grep, I see account_changed_bundle called in the link_submanifests function which my PR does not change.

castulo commented on test/functional/testlib.bash:
```diff
@@ -961,17 +961,51 @@ bump_format() {
+	sudo cp -r "$new_version_path" "$middle_version_path"
+
+	# Update +10 format and versions
+	update_manifest -p "$mom" format "$((format))"
```
I guess you could have just used "$format" here, since this is just a variable not an expression I don't think you need the (( ))

castulo commented on test/functional/testlib.bash:
```diff
@@ -961,17 +961,51 @@ bump_format() {
+		update_manifest -p "$manifest" version "$middle_version"
+		update_manifest -p "$manifest" previous "$version"
+		sudo rm -rf "$manifest".tar
+		create_tar "$manifest"
```
If you make changes to a manifest using the update_manifest function without using the _partial_ flag _-p_ the tar of the manifest will be automatically re-created. So instead of removing and creating the tar here you could just remove the paste _-p_ from the last update_manifest.

castulo commented on test/functional/testlib.bash:
```diff
@@ -1009,34 +1043,8 @@ bump_format() {
-	sudo cp "$new_version_path"/Manifest.* "$middle_version_path"/
-	sudo cp "$new_version_path"/pack* "$middle_version_path"/
-	mom="$middle_version_path"/Manifest.MoM
-	update_manifest -p "$mom" format "$((format))"
```
I see I was the one making the mistake of using the (( )) first...

castulo commented on src/manifest.c:
```diff
@@ -781,10 +783,83 @@ static bool is_installed_and_verified(struct file *file)
+
+   This implementation to create the update list does not depend on the FROM
+   file list and can be used with delta manifests. Also, when a minversion
+   upate occurs without a corresponding minversion header field in the MoM,
```
update -> update

castulo commented on test/functional/update/update-no-minversion.bats:
```diff
@@ -0,0 +1,51 @@
+
+	update_bundle "$TEST_NAME" bundle1 --add /file4
+	update_bundle "$TEST_NAME" bundle1 --update /file1
+	update_bundle "$TEST_NAME" bundle1 --delete /file2
```
to save a little bit of time you could add the -p flag to the first two update_bundle commands, the third one will take care of re-creating the tar and updating the mom anyway.

castulo commented on test/functional/update/update-no-minversion.bats:
```diff
@@ -0,0 +1,51 @@
+		Update successful. System updated from version 10 to version 20
+	EOM
+	)
+	assert_in_output "$expected_output"
```
why not _assert_is_output_ instead of _assert_in_output_? the latter will allow more output to be included, if you are not expecting that assert_is_output will make sure only the output you expect is printed.

jwakre replied:
I overlooked this, assert_is_output is leftover from the test that I copied as a starting place... All of your feedback looks good and will be addressed in the next version. Thanks!

jwakre commented:
@otaviobp @castulo I rebased my PR on top of master_new_manifests and addressed all of your feedback with one exception. The 'account_changed_bundle' function is used by link_submanifests, so I did not remove it.

otaviobp commented:
@jwakre I was talking about the account_delta_miss() and account_delta_hit(). They are out now, thanks.

castulo commented:
@jwakre I just have one last question. I see you made changes to the bump_format function in testlib so when doing a format bump the +10 version won't show all the files of a bundle as updated, only the os-release and format files show as updated. Why are you making this change? I remember when I first wrote that function that is how Tudor had told me format bump worked (or that is what I understood from my talk with him anyway). Did I get it wrong or something changed in the way format bumps are made?

jwakre commented:
> @jwakre I just have one last question. I see you made changes to the bump_format function in testlib so when doing a format bump the +10 version won't show all the files of a bundle as updated, only the os-release and format files show as updated. Why are you making this change? I remember when I first wrote that function that is how Tudor had told me format bump worked (or that is what I understood from my talk with him anyway). Did I get it wrong or something changed in the way format bumps are made?

My understanding is that the +10 and +20 should have identical content, but only the +20 should have a minversion update. The previous implementation performed a minversion update on both the +10 and +20. If I am wrong about this, I need to set the +10 minversion header field instead. @matthewrsj Can you confirm the desired format bump results?

castulo commented:
I'm ok with approving this PR as soon as @matthewrsj or @tmarcu confirm the desired result of the format bump.

With the previous version of the testlib this is how the format bumps use to look like (Manifest.os-core):

**Version 10**
```
MANIFEST	1
version:	10
previous:	0
filecount:	9
timestamp:	1538767019
contentsize:	27235

F...	a7cd26fe4b1965230c5e71e3cbc917abd4e28912737e7747b0192b1e7d77a86c	10	/core
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/lib
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr
F...	8b77bb0234b3e202a8e27bc1b4d20c9ca48582519717c8379fb58078feaa5f7a	10	/usr/lib/os-release
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/share/defaults/swupd
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/share/defaults
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/share
F...	b377bac65e56ea9dcae5e0c1f4a21faeca9c95f46d38bd8da42f99d041eab9f7	10	/usr/share/defaults/swupd/format
F...	0aa75909352ea09d1f53739b977efe5ac1065b88c9ef84199f1d2eba55dd93b9	10	/usr/share/clear/bundles/os-core
```
**Version 20**
```
MANIFEST	1
version:	20
previous:	10
filecount:	9
timestamp:	1538767021
contentsize:	27235

F...	a7cd26fe4b1965230c5e71e3cbc917abd4e28912737e7747b0192b1e7d77a86c	20	/core
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	20	/usr/lib
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	20	/usr
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	20	/usr/share/defaults/swupd
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	20	/usr/share/defaults
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	20	/usr/share
F...	0aa75909352ea09d1f53739b977efe5ac1065b88c9ef84199f1d2eba55dd93b9	20	/usr/share/clear/bundles/os-core
F...	f4debcae753a8264a99beb66d09d9be3e9d15c8687432ef3d3cbd4e470e15143	20	/usr/lib/os-release
F...	39a56dd2855a7eed2339358a2b22fe01478bbbb287e9cf13aa07bfc6714ece35	20	/usr/share/defaults/swupd/format
```
**Version 30**
```
MANIFEST	2
version:	30
previous:	20
filecount:	9
timestamp:	1538767021
contentsize:	27235

F...	a7cd26fe4b1965230c5e71e3cbc917abd4e28912737e7747b0192b1e7d77a86c	30	/core
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	30	/usr/lib
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	30	/usr
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	30	/usr/share/defaults/swupd
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	30	/usr/share/defaults
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	30	/usr/share
F...	0aa75909352ea09d1f53739b977efe5ac1065b88c9ef84199f1d2eba55dd93b9	30	/usr/share/clear/bundles/os-core
F...	f4debcae753a8264a99beb66d09d9be3e9d15c8687432ef3d3cbd4e470e15143	30	/usr/lib/os-release
F...	39a56dd2855a7eed2339358a2b22fe01478bbbb287e9cf13aa07bfc6714ece35	30	/usr/share/defaults/swupd/format
```

This is how it looks now:

**Version 10**
```
MANIFEST	1
version:	10
previous:	0
filecount:	9
timestamp:	1538767610
contentsize:	32689

F...	904912e053c3819c9af7c790b06e7725ee984cf6396eda35736985e3c3d0146b	10	/core
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/lib
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr
F...	8b77bb0234b3e202a8e27bc1b4d20c9ca48582519717c8379fb58078feaa5f7a	10	/usr/lib/os-release
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/share/defaults/swupd
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/share/defaults
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/share
F...	b377bac65e56ea9dcae5e0c1f4a21faeca9c95f46d38bd8da42f99d041eab9f7	10	/usr/share/defaults/swupd/format
F...	fa5ec26a137d185526ea4a46740fa34d8c9afb8237db8ceb65f2eb959b28f04e	10	/usr/share/clear/bundles/os-core
```
**Version 20**
```
MANIFEST	1
version:	20
previous:	10
filecount:	9
timestamp:	1538767612
contentsize:	32689

F...	904912e053c3819c9af7c790b06e7725ee984cf6396eda35736985e3c3d0146b	10	/core
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/lib
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/share/defaults/swupd
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/share/defaults
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	10	/usr/share
F...	fa5ec26a137d185526ea4a46740fa34d8c9afb8237db8ceb65f2eb959b28f04e	10	/usr/share/clear/bundles/os-core
F...	f4debcae753a8264a99beb66d09d9be3e9d15c8687432ef3d3cbd4e470e15143	20	/usr/lib/os-release
F...	39a56dd2855a7eed2339358a2b22fe01478bbbb287e9cf13aa07bfc6714ece35	20	/usr/share/defaults/swupd/format
```
**Version 30**
```
MANIFEST	2
version:	30
previous:	20
filecount:	9
timestamp:	1538767612
contentsize:	32689

F...	904912e053c3819c9af7c790b06e7725ee984cf6396eda35736985e3c3d0146b	30	/core
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	30	/usr/lib
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	30	/usr
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	30	/usr/share/defaults/swupd
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	30	/usr/share/defaults
D...	6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e	30	/usr/share
F...	fa5ec26a137d185526ea4a46740fa34d8c9afb8237db8ceb65f2eb959b28f04e	30	/usr/share/clear/bundles/os-core
F...	f4debcae753a8264a99beb66d09d9be3e9d15c8687432ef3d3cbd4e470e15143	30	/usr/lib/os-release
F...	39a56dd2855a7eed2339358a2b22fe01478bbbb287e9cf13aa07bfc6714ece35	30	/usr/share/defaults/swupd/format
```

@matthewrsj @tmarcu can any of you please confirm how it should behave?

tmarcu commented:
@castulo The output looks correct, I believe the first one was using mixer when it was incorrectly doing a minversion for **both** builds instead of just the +20 (version 30).

castulo commented:
> @castulo The output looks correct, I believe the first one was using mixer when it was incorrectly doing a minversion for **both** builds instead of just the +20 (version 30).

In that case this looks good to me
